src/com/palmergames/bukkit/towny/event/NationRankAddEvent.java

```diff
@@ -0,0 +1,65 @@
+package com.palmergames.bukkit.towny.event;
+
+import com.palmergames.bukkit.towny.object.Nation;
+import com.palmergames.bukkit.towny.object.Resident;
+import com.palmergames.bukkit.towny.object.Translation;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class NationRankAddEvent extends Event implements Cancellable {
+
+	private static final HandlerList handlers = new HandlerList();
+
+	private final Nation nation;
+	private final Resident res;
+	private final String rank;
+	private boolean isCancelled = false;
+	private String cancelMessage;
+
+	public NationRankAddEvent(Nation nation, String rank, Resident res) {
+		this.nation = nation;
+		this.rank = rank;
+		this.res = res;
+		this.cancelMessage = Translation.of("msg_resident_already_has_rank", res.getName(), "Nation");
```

Comment:
The cancelMessage should be `Translation.of("msg_resident_not_part_of_any_town")` which can be set directly on line 18.